pkg/cloud/services/autoscaling/autoscalinggroup_test.go

```diff
@@ -220,3 +317,722 @@ func TestService_SDKToAutoScalingGroup(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	tests := []struct {
+		name            string
+		machinePoolName string
+		expect          func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder)
+		check           func(out bool, err error)
+	}{
+		{
+			name:            "should return error if describe instance refresh failed",
+			machinePoolName: "mpn",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				m.DescribeInstanceRefreshes(gomock.Eq(&autoscaling.DescribeInstanceRefreshesInput{
+					AutoScalingGroupName: aws.String("mpn"),
+				})).
+					Return(nil, awserrors.NewNotFound("not found"))
+			},
+			check: func(out bool, err error) {
+				if err == nil {
+					t.Fatalf("expected an error")
+				}
+				if out {
+					t.Fatalf("expected to be false")
+				}
+			},
+		},
+		{
+			name:            "should return true if no instance available for refresh",
+			machinePoolName: "mpn",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				m.DescribeInstanceRefreshes(gomock.Eq(&autoscaling.DescribeInstanceRefreshesInput{
+					AutoScalingGroupName: aws.String("mpn"),
+				})).
+					Return(&autoscaling.DescribeInstanceRefreshesOutput{}, nil)
+			},
+			check: func(out bool, err error) {
+				if err != nil {
+					t.Fatalf("did not expect an error: %+v", err)
+				}
+				if !out {
+					t.Fatalf("expected to be true")
+				}
+			},
+		},
+		{
+			name:            "should return false if some instances have unfinished refresh",
+			machinePoolName: "mpn",
+			expect: func(s *mock_cloudiface.MockClusterScoperMockRecorder, m *mock_autoscalingiface.MockAutoScalingAPIMockRecorder) {
+				m.DescribeInstanceRefreshes(gomock.Eq(&autoscaling.DescribeInstanceRefreshesInput{
+					AutoScalingGroupName: aws.String("mpn"),
+				})).
+					Return(&autoscaling.DescribeInstanceRefreshesOutput{
+						InstanceRefreshes: []*autoscaling.InstanceRefresh{
+							{
+								Status: aws.String(autoscaling.InstanceRefreshStatusInProgress),
+							},
+						},
+					}, nil)
+			},
+			check: func(out bool, err error) {
+				if err != nil {
+					t.Fatalf("did not expect an error: %+v", err)
+				}
+				if out {
+					t.Fatalf("expected to be false")
+				}
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			asgMock := mock_autoscalingiface.NewMockAutoScalingAPI(mockCtrl)
+			scopeMock := mock_cloudiface.NewMockClusterScoper(mockCtrl)
+			tt.expect(scopeMock.EXPECT(), asgMock.EXPECT())
+			s := &Service{
+				scope:     scopeMock,
+				ASGClient: asgMock,
+			}
+			mps := &scope.MachinePoolScope{
+				AWSMachinePool: &expinfrav1.AWSMachinePool{},
+			}
+			mps.AWSMachinePool.Name = tt.machinePoolName
+			out, err := s.CanStartASGInstanceRefresh(mps)
+			tt.check(out, err)
+		})
+	}
+}
+
```

Comment:
Can we also add tests for `StartASGInstanceRefresh` func?
Also, can you check if `createSDKMixedInstancesPolicy` func can be covered in any of the flows?